src/Core/Checkout/Promotion/Subscriber/PromotionIndividualCodeRedeemer.php

```diff
@@ -25,21 +32,54 @@ class PromotionIndividualCodeRedeemer implements EventSubscriberInterface
      * @internal
      *
      * @param EntityRepository<PromotionIndividualCodeCollection> $codesRepository
+     * @param EntityRepository<OrderCollection> $orderRepository
      */
-    public function __construct(private readonly EntityRepository $codesRepository)
-    {
+    public function __construct(
+        private readonly EntityRepository $codesRepository,
+        private readonly EntityRepository $orderRepository
+    ) {
     }
 
     public static function getSubscribedEvents(): array
     {
         return [
-            CheckoutOrderPlacedEvent::class => 'onOrderPlaced',
+            CheckoutOrderPlacedEvent::class => 'onOrderProcess',
+            OrderEvents::ORDER_LINE_ITEM_WRITTEN_EVENT => 'onOrderLineItemWritten',
         ];
     }
 
-    public function onOrderPlaced(CheckoutOrderPlacedEvent $event): void
+    public function onOrderLineItemWritten(EntityWrittenEvent $event): void
+    {
+        if ($event->getContext()->getVersionId() !== Defaults::LIVE_VERSION) {
+            return;
+        }
+
+        $orderLineItems = new OrderLineItemCollection();
+
+        foreach ($event->getWriteResults() as $result) {
+            $orderLineItems->add((new OrderLineItemEntity())->assign($result->getPayload()));
+        }
+
+        $lineItem = $orderLineItems->first();
+        \assert($lineItem instanceof OrderLineItemEntity);
+
+        $order = $this->orderRepository->search(
+            new Criteria([$lineItem->getOrderId()]),
+            $event->getContext()
+        )->getEntities()->first();
+        \assert($order instanceof OrderEntity);
```

Comment:
I don't like fetching the whole order. As far as I can see, we only need the OrderCustomer, right? (which you are by the way not adding to your Criteria as an association here).